dts/bindings/radio_fem/skyworks,sky66112-11.yaml

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+description: >
+    This is a representation of the Skyworks SKY66112-11 Radio Front-End module.
+
+compatible: "skyworks,sky66112-11"
+
+include: base.yaml
+
+properties:
+    ctx-gpios:
+        type: phandle-array
+        required: false
+        description: >
+            Gpio of the SOC controlling CTX pin of the SKY66112-11
+
+    crx-gpios:
+        type: phandle-array
+        required: false
+        description: >
+            Gpio of the SOC controlling CRX pin of the SKY66112-11
+
+    ctx-settle-time-us:
+        type: int
+        default: 23
+        description: >
+            Settling time in microseconds from activation of CTX to transmit.
+
+            The default value has been determined experimentally: it is the
+            minimum time interval between asserting the CTX signal and starting
+            the radio transmission, which does not generate unacceptable
+            interference level on adjacent channels. The tests concerned a
+            configuration in which nRF5x SoC controlling the SKY66112-11.
+            The value can be adjusted for other tuned configurations.
+
+    crx-settle-time-us:
+        type: int
+        default: 5
+        description: >
+            Settling time in microseconds from activation of CRX to receive.
+
+            The default value has been determined experimentally: it is the
+            minimum time interval between asserting the CRX signal and starting
+            the radio reception, which does not degrade transmission quality
+            indicators (compared to a constantly turned on FEM LNA). The tests
+            concerned a configuration in which nRF5x SoC controlling
+            the SKY66112-11. The value can be adjusted for other tuned
+            configurations.
+
+    tx-gain-db:
+        type: int
+        default: 22
+        description: >
+            TX gain of the PA amplifier of SKY66112-11 in dB.
+
+            Default value is based on Table 5 of the SKY66112-11 data sheet
+            (July 19, 2019). This is a fixed parameter of the device. The value
+            can be changed if, instead of SKY66112-11, another device with
+            different gain, but compatible with the control interface via the
+            CTX, CRX pins is used.
+
+    rx-gain-db:
+        type: int
+        default: 11
+        description: >
+            RX gain of the LNA amplifier of SKY66112-11 in dB.
+
+            Default value is based on Table 5 of the SKY66112-11 data sheet
+            (July 19, 2019). This is a fixed parameter of the device. The value
+            can be changed if, instead of SKY66112-11, another device with
+            different gain, but compatible with the control interface via the
+            CTX, CRX pins is used.
```

Comment:
If these are truly fixed parameters of the device, I don't follow why they are here in the binding. The driver should be able to `#define` them based on the compatible itself. If you want to use "another device", that would require its own devicetree binding.

Reply:
The driver for sky66112 can be used with multiple similar front end modules. The modules differ only in required timings and gain values. It would be hard to maintain and to create another very similar binding and driver for each of nearly identical devices, that differs only in one `#define` (and potentially some timings defined in DTS).

What would be the correct way to provide a generic driver and matching binding, that can be used with sky66112 and similar front end modules?

Reply:
I have no experience with FEMs, but in general there are two options when defining bindings for multiple related devices:

1. create a YAML "include" file that does not define a compatible but defines the common properties, and include it from the "real" binding files for each actual device. An example "include" file is nordic,nrf-spi-common.yaml, which is included by the "real" binding files nordic,nrf-spim.yaml and nordic,nrf-spi.yaml.

2. create a generic binding that describes the entire category of devices and is supported by a common driver. I think jedec,spi-nor.yaml is an example.

Reply:
Let my explain our intentions here. We would like to add to NCS support for Front-End Modules like:

- SKY66112
- SKY66113 (using subset of features)
- SKY66114
- SKY66403
- SKY66404
- 8TR8201

Each of these devices have the same control interface. However, they differ in gain values and slightly in required timings. These are just examples, I don't think that I've found all FEMs sharing this simple generic GPIO interface. Becasue of that we would like to make this list as easily extensible as possible. I don't think that this simple interface (two GPIO pins, one used to indicate TX, another one used to indicate RX) is anywhere specified like Jedec SPI. Moreover, these devices sharing the same simple control interface are created by multiple vendors.

The driver for SKY66112 can be easily used for controlling other devices. Usually timings we have for SKY66112 apply to other. The only thing that should be adjusted is gain of particular device in given design (please note that gain may vary for the same device depending on RF path design on given board).

Our goal is to enable NCS users, that have one of FEMs mentioned above or similar in their design, to easily enable FEM support in NCS project. The easiest way is to define control pin numbers (done in device tree), define gains (if differ from the example we provided), and optionally tweak timings (usually not necessary).

I'm not sure which of the options you suggested applies to our intentions.

Reply:
Thanks for explaining. @AdamZelikNS and I discussed offline and I think we settled on using a generic include file with individual hardware bindings for real devices that mpsl_fem.c finds in the tree is the way to go.

We also discussed using a chosen node or an alias as a more flexible way to find the desired node. Adam will be looking into it further.

Reply:
I came to the conclusion that using alias or chosen node not be of benefit in this case. This is due to the fact that there is most probably only one FEM instance in the system.

The approach that, as I applied, i.e. 
  - there is generic-fem-two-ctrl-pins DT binding
  - binding for SKY66112 supplements the generics one with device-specific values
  - driver (and its Kconfig) is generic as it are looking for `compatible = "generic-fem-two-ctrl-pins"` and using it

is well suited to our needs.